highfidelity/hifi
# Trying to fix QtQml/QtScript crashes

This PR attempts to resolve the increased frequency of bugsplat crashes in QtQml and QtScript.  I believe that the primary issue is race conditions between the scripting thread which might be inadvertently calling QML functions on the script thread, corrupting memory and resulting in these crashes deep inside the Qt DLLs.  

* Move `ToolbarScriptingInterface` from `interface` to `ui`
* Move `TabletScriptingInterface` from `script-engine` to `ui`
* Remove circular dependency between `ui` and `script-engine` (in fact both directions have been removed) 
* Remove dependency from `networking` to `QtWebEngine` and all `QtWebEngine` dependant types from `networking` to `ui` (fixes bug in building server only components on systems with no QtWebEngine library installed)
* Migrate `ToolbarScriptingInterface` and `TabletScriptingInterface` to to rely on the main thread
  * Forced all code that might be called from a script in these interfaces should use Qt method invocation to execute on the main thread.  All other code should verify it's running on the main thread (mostly by Q_ASSERT calls right now)
  * All proxy objects returned by these interfaces should be owned by the main thread.  Will emit warnings if not
  * Removed all mutexes from `TabletScriptingInterface` and related proxy classes in favor of forcing execution on the main thread
  * Similary, destructors for the proxy classes verify that they are running on the main thread, or log warnings
* Fix a bug in `RunningScripts.qml` that triggers a heap corruption warning Visual Studio

## Likely related fogbugs

[5535 CRASH in QtScript](https://highfidelity.fogbugz.com/f/cases/edit/5535/CRASH-in-QtScript)
[5518 CRASH in QtQml](https://highfidelity.fogbugz.com/f/cases/5518/CRASH-in-QtQml-maybe-in-Steam)

## Testing

* Verify functionality of scripts that manipulate toolbars / tablets.  
* Verify that the behavior is correct (well, unchanged anyway) for both tablet and toolbar modes when using the tablet interface.  
* Verify that the button behavior is correct when _switching_ between tablet and toolbar mode for a script that uses the tablet interface

I did some testing with [this script](https://wagasi.tcmeta.net/hifi/testing/wardrobe/Wardrobe.js) which I noted in one of the crash logs.  I've loaded and unloaded it, as well as tried shutting down and restarting with it running.  I've seen at least one crash doing this, as well as some mysterious exceptions that get throw and caught deep inside QtQml, though I'm not sure if the latter is related or an actual issue.  However, testing should be done using this script or another like it that exercises the tablet scripting interface to see if a crash can be reproduced both on master and with this PR.  I hope that this PR should be more stable.
